### ISSUE TITLE
Navigation buttons to move left and right on a chart/graph

### ISSUE THREAD

**Issue by abhijit945** [cerner/carbon-graphs#41]
There were three options for Panning in Carbon, each with pros and cons: 

1. Browser default scroll bar
- this is the browser default scroll bar because there is too much data to show
- would appear if we turn off Carbon resize
- there is no return button option because all data is loaded
- the consumer decides what the look back is and what “all data” means (the look back fetches more data, for example 1mo, 1 year, by consumer control)
- this isn’t carbon, isn’t really panning, this is all browser
- it would be arrow based navigation for keyboard accessibility
- would work for any graph type

2. Return/left/right button control
- Return/left/right button control
- Doesn’t load all data in advance
- Carbon will show loading while the consumer feeds more data
- consumer provides answers each interaction to if there is more data outside of the arrows. we can dither the arrow if there isn’t more data outside the view. carbon will save the state of initial load and return will take them there.
- We won’t disable resize. 
- Works for multiple graph types
- Would be tab to navigate, enter to select
- Text on the return button would require each consumer to define, or Carbon could do an icon and English-only Tooltip
- Works better with zooming buttons 

3. D3 panning
- All data must be loaded in advance
- no keyboard navigation
- no indication you can pan
- no return button option
- Would disable resize because it is a requirement from d3 to use panning
- only works for line and bar
- We would have to disable zoom out
- https://bl.ocks.org/mbostock/7ec977c95910dd026812

We cannot make our own scroll bar (browser is the only option) and we cannot have #2 or #3 together because we won’t know what a “lower step” is for each of the left/right buttons because everything is loaded. Ultimately, to be the most accessible, we will move forward with only left and right arrows to move by each lower step increments because panning from D3 offers no affordance you can click and drag. We will not have a return button at the moment because text is not well supported in this environment and there is not a good icon for return that is as universally understood as the navigation arrows. We can also dither the arrows based on knowing if there is more data to either side of the visible graph. 

They will be placed in the upper right-hand corner of the graph aligning with edge of the y-axis. 

![image](https://user-images.githubusercontent.com/4908886/61049991-88b7e900-a3ab-11e9-8512-4364c5d0e0ae.png)

![image](https://user-images.githubusercontent.com/4908886/61050027-a08f6d00-a3ab-11e9-866e-dfa1d3ae4b07.png)

![image](https://user-images.githubusercontent.com/4908886/61050034-a4bb8a80-a3ab-11e9-835f-933ab7d3ed3b.png)

**Comment by abhijit945** [comment]
**Note to self**: `PaginationControl`
```javascript
{
	PREVIOUS: {
		key: "uid_previous",
		label: {
			// This can be locale based, by default this will be blank
			display: "Previous"
		}
		shape: {
			// Left arrow
		},
		isPreviousDisabled: false,
		previousClickEventHandler: () => {}
	},
	NEXT: {
		key: "uid_next",
		label: {
			// This can be locale based, by default this will be blank
			display: "Next"
		}
		shape: {
			// Right arrow
		},
		isNextDisabled: false,
		nextClickEventHandler: () => {}
	}
}
```

For "loading screen" we are opting for `SVG feGaussianBlur`. 

https://bl.ocks.org/mbostock/1342359

Consumer input:
```
pan: {
    enabled: true,
    showControls: true
}
```

Instead of peppering code with conditionals, we could have a dictionary `subconfig` of settings that we need to enable/disable such as `clamp` and based on those properties we will set the values. 

Before: 
```
scale.x = d3.time
        .scale()
        .domain(config.axis.x.domain)
        .range([0, getXAxisWidth(config)])
        .clamp(true);
```
After: 
```
// Create a dictionary
isPanningModeEnabled()
        ? {
              shouldClamp: false
          }
        : {
              shouldClamp: true
          };
```
```
// In axis.js
scale.x = d3.time
	.scale()
	.domain(config.axis.x.domain)
	.range([0, getXAxisWidth(config)])
	.clamp(consolidatedDictionary.shouldClamp);
```

**Comment by mrw100** [comment]
We have a similar need to disable clamping, but don't need the navigation buttons because the ability to navigate will be from terra-timeline. Ideally, we would like an option to be able to disable that functionality similar to the above isPanningModeEnabled option.  Should we log a separate issue for this functionality or can our need be considered with this issue?

**Comment by abhijit945** [comment]
The controls and the mode dont need to be on the same PR but they need to support each other for consumers who dont need a timeline based panning or dont need all the functionality within timeline framework. 

@Prashant-Kamboj was planning to work towards this, not sure how much progress he has done. Didnt  you have a prototype already for this?

**Comment by Prashant-Kamboj** [comment]
I am working on this implementation and able to achieve some work is still pending for bar graph and the dateline indicator inclusion also.

**Comment by erinplourde** [comment]
Here are links to the arrow icons: 
Left - https://github.com/cerner/one-cerner-style-icons/blob/master/src/icons/0026_chevronLeft_a.svg
Right - https://github.com/cerner/one-cerner-style-icons/blob/master/src/icons/0026_chevronRight_a.svg

**Comment by laurunning** [comment]
Does this issue include the data point running off the edge of the viewable area?

**Comment by Matthematic** [comment]
@ryanthemanuel  @abhijit945  Can someone ELI5 how a graph under the stella-timeline would pan in response to navigation events? The PR above for adding the panning property added a demo into the site, but it looks like that demo is still destroying/remaking the graph upon navigation as opposed to actual panning.

**Comment by laurenbock** [comment]
The contribution is only disabling the animations and not the complete panning functionality. The full functionality of smooth panning is still being developed and is a major change.

**Comment by abhijit945** [comment]
Duplicate of https://github.com/cerner/carbon-graphs/issues/93
Closing this since more recent activity and comments are on that issue.
